proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2948,198 +3035,256 @@ lemma cancelIPC_ccorres1:
                    (UNIV \<inter> {s. tptr_' s = tcb_ptr_to_ctcb_ptr thread}) []
           (cancelIPC thread) (Call cancelIPC_'proc)"
   apply (cinit lift: tptr_' simp: Let_def cong: call_ignore_cong)
-sorry (* FIXME RT: cancelIPC_ccorres1 *) (*
    apply (rule ccorres_move_c_guard_tcb)
-   apply csymbr
+   apply (rule ccorres_stateAssert)
+   apply (rule ccorres_stateAssert)
    apply (rule getThreadState_ccorres_foo)
-   apply (rule ccorres_symb_exec_r)
-     apply (rule_tac xf'=ret__unsigned_longlong_' in ccorres_abstract, ceqv)
-     apply (rule_tac P="rv' = thread_state_to_tsType rv" in ccorres_gen_asm2)
-     apply wpc
-            \<comment> \<open>BlockedOnReceive\<close>
-            apply (simp add: word_sle_def ccorres_cond_iffs cong: call_ignore_cong)
-            apply (rule ccorres_rhs_assoc)+
-            apply csymbr
-            apply csymbr
-            apply (rule ccorres_pre_getEndpoint)
-            apply (rule ccorres_assert)
-            apply (rule ccorres_symb_exec_r) \<comment> \<open>ptr_get lemmas don't work so well :(\<close>
+   apply (rename_tac threadState)
+   apply csymbr
+   apply csymbr
+   apply (rule ccorres_split_nothrow)
+       apply (rule threadSet_ccorres_lemma2[where P=\<top>])
+        apply vcg
+       apply (clarsimp simp: typ_heap_simps')
+       apply (erule(1) rf_sr_tcb_update_no_queue2, (simp add: typ_heap_simps')+)[1]
+        apply (rule ball_tcb_cte_casesI, simp_all)[1]
```

Comment:
this might be copied from and old proof, these days we can do
`apply (rule ball_tcb_cte_casesI; simp)` instead

Also, the `erule(1)` above it should be `erule (1)`, so this might be ancient, since that line probably could also be rewritten as:
`apply (erule (1) rf_sr_tcb_update_no_queue2; simp add: typ_heap_simps')`

Reply:
I made that first change, but the second simplification does not work.

Reply:
Well, it doesn't always work out. `;` processes new goals last-to-first so there might be a unification issue. Fine to leave that as is then.